### PR TITLE
[mpeg2d] Invalid slices handling

### PR DESCRIPTION
**Pull request by asmirno1**
1) Handling case when vertical_position of slice is out of range
2) Recalculating numberMBsInSlice of prev slice if two
slices are located at the same row

### CODE REVIEW

**Comment by asmirno1** [comment]
> I don't see any call of EliminateSliceErrors function ...

Missed this part. Fixed.

File: _studio/shared/umc/codec/mpeg2_dec/include/umc_mpeg2_decoder.h [hunk 1]
```diff
@@ -187,6 +187,7 @@ namespace UMC_MPEG2_DECODER
 
         // Action on new picture
         virtual bool OnNewPicture(); // returns true on full frame
+        virtual UMC::Status EliminateSliceErrors(MPEG2DecoderFrame& frame, uint8_t fieldIndex);
```
Comment (dmitryermilov) [inline _studio/shared/umc/codec/mpeg2_dec/include/umc_mpeg2_decoder.h]: Return status doesn't seem to be needed here

File: _studio/shared/umc/codec/mpeg2_dec/src/umc_mpeg2_bitstream.cpp [hunk 2]
```diff
@@ -295,6 +296,15 @@ namespace UMC_MPEG2_DECODER
         if (CheckBSLeft())
             throw mpeg2_exception(UMC::UMC_ERR_INVALID_STREAM);
 
+        uint32_t width_in_MBs = (uint32_t)(mfx::align2_value<uint32_t>(slice.GetSeqHeader().horizontal_size_value, 16) / 16);
```
Comment (mgonchar) [inline _studio/shared/umc/codec/mpeg2_dec/src/umc_mpeg2_bitstream.cpp]: I think explicit cast is not needed, just use unsigned literals:
```
uint32_t width_in_MBs = mfx::align2_value<uint32_t>(slice.GetSeqHeader().horizontal_size_value, 16) / 16u;
```

Comment (asmirno1) [reply _studio/shared/umc/codec/mpeg2_dec/src/umc_mpeg2_bitstream.cpp]: Done

File: _studio/shared/umc/codec/mpeg2_dec/src/umc_mpeg2_decoder.cpp [hunk 3]
```diff
@@ -903,6 +903,32 @@ namespace UMC_MPEG2_DECODER
...
+            auto slice = frameInfo.GetSlice(sliceNum);
+            auto sliceHeader = slice->GetSliceHeader();
+
+            // Slice may cover just a part of a row (not full row), this is indicated by macroblockAddressIncrement
+            // Looping over slices, we heed to check if next slice has macroblockAddressIncrement > 0,
+            // this means that we need to recalculate numberMBsInSlice for previous slice
+            if (sliceNum > 0 && sliceHeader.macroblockAddressIncrement > 0)
+            {
+                auto prevSlice = frameInfo.GetSlice(sliceNum - 1); //take previous
+                auto& prevSliceHeader = prevSlice->GetSliceHeader();
```
Comment (asmirno1) [inline _studio/shared/umc/codec/mpeg2_dec/src/umc_mpeg2_decoder.cpp]: Added "&" in new patch set. Previously, without it, a copy of prevSliceHeader was created, but we need to change prevSliceHeader.numberMBsInSlice value in line 926. So now it is changing correctly.

**Comment by uartie** [comment]
Does the buildbot/test include mpeg2 decode cases?  Because this patchset breaks every mpeg2 decode case in https://github.com/intel/vaapi-fits.

**Comment by uartie** [comment]
The regression caused by this patch is clearly visible with the following command:
```
$ GST_VAAPI_ALL_DRIVERS=1 LIBVA_DRIVER_NAME=iHD gst-launch-1.0 -vf \
  videotestsrc horizontal-speed=2 ! msdkmpeg2enc ! msdkmpeg2dec ! vaapisink
```

**Comment by uartie** [comment]
artifacts

**Comment by uartie** [comment]
regression

**Comment by dmitryermilov** [comment]
Thank you. And thanks again for reporting it so fast:)

File: _studio/shared/umc/codec/mpeg2_dec/src/umc_mpeg2_bitstream.cpp [hunk 4]
```diff
@@ -295,6 +296,15 @@ namespace UMC_MPEG2_DECODER
         if (CheckBSLeft())
             throw mpeg2_exception(UMC::UMC_ERR_INVALID_STREAM);
 
+        uint32_t width_in_MBs  = mfx::align2_value<uint32_t>(slice.GetSeqHeader().horizontal_size_value, 16) / 16u;
+        uint32_t height_in_MBs = mfx::align2_value<uint32_t>(slice.GetSeqHeader().vertical_size_value, 16) / 16u;
+
+        // invalid slice - skipping it
+        if (slice.sliceHeader.slice_vertical_position >= height_in_MBs)
```
Comment (uartie) [inline _studio/shared/umc/codec/mpeg2_dec/src/umc_mpeg2_bitstream.cpp]: Should this be `if (slice.sliceHeader.slice_vertical_position > height_in_MBs)` instead?  It seems to get rid of the regression I reported.

Comment (asmirno1) [reply _studio/shared/umc/codec/mpeg2_dec/src/umc_mpeg2_bitstream.cpp]: Yep, I confirm that's a bug.

**Pull request merged**